crates/kitsune_p2p/fetch/src/pool.rs

```diff
@@ -458,25 +470,120 @@ mod tests {
+        let mut sources = Sources(vec![
+            SourceRecord {
+                source: source(1),
+                last_request: Some(Instant::now()), // recently tried
+            }
+            .into(),
+            SourceRecord {
+                source: source(2),
+                last_request: None, // never checked
+            }
+            .into(),
+            SourceRecord {
+                source: source(3),
+                last_request: None, // never checked
+            }
+            .into(),
+        ]);
+
+        assert_eq!(sources.next(source_delay), Some(source(2)));
+
+        // All sources now past their retry delay
+        tokio::time::advance(source_delay).await;
+
+        // Would expect source 1 to be tried next but trying 2 first rotated 1 and 2 to the end of the list
+        assert_eq!(sources.next(source_delay), Some(source(3)));
+        assert_eq!(sources.next(source_delay), Some(source(1)));
+        assert_eq!(sources.next(source_delay), Some(source(2)));
+        assert_eq!(sources.next(source_delay), None);
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn source_rotation_uses_all_sources() {
+        let mut noise = [0; 1_000];
+        rand::thread_rng().fill_bytes(&mut noise);
+        let mut u = Unstructured::new(&noise);
+
+        let source_delay = Duration::from_secs(rand::thread_rng().gen_range(1..50));
+
+        let v = std::iter::repeat_with(|| Duration::arbitrary(&mut u).unwrap())
+            .take(100)
+            .enumerate()
+            .map(|(i, duration)| SourceRecord {
+                source: source(i as u8),
+                last_request: Instant::now().checked_sub(duration),
+            })
+            .collect();
+        let mut sources = Sources(v);
+
+        let mut seen_sources: HashSet<u8> = HashSet::new();
+        for _ in 0..100 {
+            if let Some(s) = sources.next(source_delay) {
+                match s {
+                    FetchSource::Agent(a) => {
+                        // The source agent key is a repeating byte array of the source number, so we can retrieve any
+                        // byte here to get the source number
+                        seen_sources.insert(a.0[0]);
+                    }
+                }
+            }
+
+            tokio::time::advance(source_delay).await;
+        }
+
+        assert_eq!(100, seen_sources.len());
+    }
+
+    #[test]
+    fn state_removes_context_on_merge_if_new_is_none() {
+        let mut q = State::default();
+        let cfg = Config(1, 1);
+
+        q.push(&cfg, req(1, ctx(1), source(1)));
+        assert_eq!(ctx(1), q.queue.front().unwrap().1.context);
+
+        // Same key but difference source so that it will merge, but wipe out the context
+        q.push(&cfg, req(1, None, source(0)));
+        assert_eq!(None, q.queue.front().unwrap().1.context);
+    }
+
+    #[test]
+    fn state_ignores_duplicate_sources_on_merge() {
```

Comment:
This test fails, we are allowed to add the same source twice and because (as commented above) sources are independent, this code will permit sending duplicate requests to the same host.

I haven't not verified whether this ever happens in reality but it is a detail that this code should be responsible for deciding how to handle. Either by rejecting the `push` into the fetch queue or by quietly ignoring or replacing duplicates. Undecided which option is best and some input would be useful here

Reply:
Yah... seems like `FetchPoolItem.sources` also needs to be based off `LinkedHashMap` since we're bothering to import that anyways.

Reply:
The key would be the `FetchSource`.

Reply:
Yep, works nicely. I've made the change https://github.com/holochain/holochain/pull/2643/commits/e63d50eb0e1c256a6f9a099e5dbf441f2a366c57

It does come with a subtle behavior change where new sources are added to the BACK of the 'list' not the FRONT like they used to be. That caused the `queue_push` test to need to change.

I don't see this as a big problem because existing sources will be on timeout most of the time anyway so the next iteration over `Sources` will pick the new source anyway.

If that was intentional to try new sources first then I can fix the ordering by running through the linked map but it doesn't have an `insert_front` or similar unfortunately. That isn't likely to change because the library is in maintenance https://github.com/contain-rs/linked-hash-map.

Reply:
I think this is good as-is :+1:

Reply:
@ThetaSinner - as a side note (and likely a lower priority completely different PR) - I wonder if we should switch to a more maintained library - We're already importing [indexmap](https://crates.io/crates/indexmap/2.0.0) through our serde_json usage.

Reply:
Yes that's a good idea and that option one of the alternatives I came across while taking a look around. I'll open a ticket for it